ca/certificate-authority.go

```diff
@@ -216,6 +272,57 @@ func (ca *CertificateAuthorityImpl) noteHSMFault(err error) {
 	return
 }
 
+// Extract supported extensions from a CSR.  The following extensions are
+// currently supported:
+//
+// * 1.3.6.1.5.5.7.1.24 - TLS Feature [RFC7633], with the "must staple" value.
+//                        Any other value will result in an error.
+//
+// Other requested extensions are silently ignored.
+func (ca *CertificateAuthorityImpl) extensionsFromCSR(csr *x509.CertificateRequest) (extensions []signer.Extension, err error) {
+	extensions = []signer.Extension{}
+	for _, attr := range csr.Attributes {
+		if !attr.Type.Equal(oidExtensionRequest) {
+			continue
+		}
+
+		for _, extList := range attr.Value {
+			for _, ext := range extList {
+				switch {
+				case ext.Type.Equal(oidTLSFeature):
+					ca.stats.Inc(metricCSRExtensionTLSFeature, 1, 1.0)
+					value, ok := ext.Value.([]byte)
+					if !ok {
+						msg := fmt.Sprintf("Mal-formed extension with OID %v", ext.Type)
+						err = core.CertificateIssuanceError(msg)
+						return
+					} else if !bytes.Equal(value, mustStapleFeatureValue) {
+						msg := fmt.Sprintf("Unsupported value for extension with OID %v", ext.Type)
+						ca.stats.Inc(metricCSRExtensionTLSFeatureInvalid, 1, 1.0)
+						err = core.CertificateIssuanceError(msg)
+						return
+					}
+
+					extensions = append(extensions, mustStapleExtension)
```

Comment:
Mainly what I was worried about is that this would allow a user to create
arbitrarily sized certificates by just sending a massive CSR with a bunch
of duplicates. I haven't tested it but the `x509.CertificateRequest` type
seems to make this quite easy to create.

On Mon, Feb 1, 2016 at 1:50 PM bifurcation <notifications@github.com> wrote:

> In ca/certificate-authority.go
> <https://github.com/letsencrypt/boulder/pull/1224#discussion_r51486553>:
>
> > +				switch {
> > +				case ext.Type.Equal(oidTLSFeature):
> > +					ca.stats.Inc(metricCSRExtensionTLSFeature, 1, 1.0)
> > +					value, ok := ext.Value.([]byte)
> > +					if !ok {
> > +						msg := fmt.Sprintf("Mal-formed extension with OID %v", ext.Type)
> > +						err = core.CertificateIssuanceError(msg)
> > +						return
> > +					} else if !bytes.Equal(value, mustStapleFeatureValue) {
> > +						msg := fmt.Sprintf("Unsupported value for extension with OID %v", ext.Type)
> > +						ca.stats.Inc(metricCSRExtensionTLSFeatureInvalid, 1, 1.0)
> > +						err = core.CertificateIssuanceError(msg)
> > +						return
> > +					}
> > +
> > +					extensions = append(extensions, mustStapleExtension)
>
> I don't think we need to support duplicate extensions, but I'm not that
> worried about forbidding it either. In practice, I don't expect this to be
> an issue. I don't even know how to create a CSR with duplicate extensions
> requested. Even if someone does figure out how to do it, having multiple
> copies of this one shouldn't be a problem. Maybe just add a comment for
> now, and if we support more extensions in the future, then we should
> address it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/letsencrypt/boulder/pull/1224/files#r51486553>.
>